### PR TITLE
workflow components to apply follow-up of background / zerolag and rerank

### PR DESCRIPTION
**Pull request by ahnitz**
This PR provides the workflow support to re-rank candidates using a follow-up script.

### CODE REVIEW

**Comment by ahnitz** [comment]
Example output if you don't provide the new "workflow-rerank" section on a multi-ifo ini file. It should skip it and behave as normal. You will get an additional message in the log however. 
```
2019-07-01 13:27:23,592:INFO : Overriding section results_page option output-path with value ../../../html in configuration.
2019-07-01 13:27:23,645:INFO : Overriding section results_page option output-path with value ../../../html in configuration.
2019-07-01 13:27:23,729:INFO : Overriding section results_page option output-path with value ../../../html in configuration.
2019-07-01 13:27:23,759:INFO : Created log file /work/ahnitz/projects/2-ogc/t5/O2/98/output/results/8._workflow/H1L1V1-WORKFLOW-LOG-1186896976-44766.txt
Downloading https://www.gw-openscience.org/timeline/segments/json/O2_16KHZ_R1/H1_DATA/1186896976/44766/
|======================================================================================================| 123 /123  (100.00%)         0s
Downloading https://www.gw-openscience.org/timeline/segments/json/O2_16KHZ_R1/H1_DATA/1186896976/44766/
|======================================================================================================| 123 /123  (100.00%)         0s
Downloading https://www.gw-openscience.org/timeline/segments/json/O2_16KHZ_R1/H1_CBC_CAT1/1186896976/44766/
|======================================================================================================| 127 /127  (100.00%)         0s
2019-07-01 13:27:26,204:INFO : H1: got segments-science flags
Downloading https://www.gw-openscience.org/timeline/segments/json/O2_16KHZ_R1/L1_DATA/1186896976/44766/
|======================================================================================================| 123 /123  (100.00%)         0s
Downloading https://www.gw-openscience.org/timeline/segments/json/O2_16KHZ_R1/L1_DATA/1186896976/44766/
|======================================================================================================| 123 /123  (100.00%)         0s
Downloading https://www.gw-openscience.org/timeline/segments/json/O2_16KHZ_R1/L1_CBC_CAT1/1186896976/44766/
|======================================================================================================| 127 /127  (100.00%)         0s
2019-07-01 13:27:28,189:INFO : L1: got segments-science flags
Downloading https://www.gw-openscience.org/timeline/segments/json/O2_16KHZ_R1/V1_DATA/1186896976/44766/
|======================================================================================================| 149 /149  (100.00%)         0s
Downloading https://www.gw-openscience.org/timeline/segments/json/O2_16KHZ_R1/V1_DATA/1186896976/44766/
|======================================================================================================| 149 /149  (100.00%)         0s
Downloading https://www.gw-openscience.org/timeline/segments/json/O2_16KHZ_R1/V1_CBC_CAT1/1186896976/44766/
|======================================================================================================| 153 /153  (100.00%)         0s
2019-07-01 13:27:30,174:INFO : V1: got segments-science flags
2019-07-01 13:27:30,183:INFO : Entering datafind module
2019-07-01 13:27:30,184:INFO : Starting datafind with setup_datafind_runtime_generated
2019-07-01 13:27:30,184:INFO : Setting up connection to datafind server.
2019-07-01 13:27:30,184:INFO : Querying datafind server for all science segments.
2019-07-01 13:27:30,234:INFO : setup_datafind_runtime_generated completed
2019-07-01 13:27:30,243:INFO : Leaving datafind module
Downloading https://www.gw-openscience.org/timeline/segments/json/O2_16KHZ_R1/H1_DATA/1186896976/44766/
|======================================================================================================| 123 /123  (100.00%)         0s
Downloading https://www.gw-openscience.org/timeline/segments/json/O2_16KHZ_R1/H1_CBC_CAT2/1186896976/44766/
|======================================================================================================| 127 /127  (100.00%)         0s
Downloading https://www.gw-openscience.org/timeline/segments/json/O2_16KHZ_R1/H1_DATA/1186896976/44766/
|======================================================================================================| 123 /123  (100.00%)         0s
Downloading https://www.gw-openscience.org/timeline/segments/json/O2_16KHZ_R1/H1_NO_CBC_HW_INJ/1186896976/44766/
|======================================================================================================| 132 /132  (100.00%)         0s
Downloading https://www.gw-openscience.org/timeline/segments/json/O2_16KHZ_R1/H1_DATA/1186896976/44766/
|======================================================================================================| 123 /123  (100.00%)         0s
Downloading https://www.gw-openscience.org/timeline/segments/json/O2_16KHZ_R1/H1_NO_BURST_HW_INJ/1186896976/44766/
|======================================================================================================| 134 /134  (100.00%)         0s
2019-07-01 13:27:34,255:INFO : H1: got segments-vetoes flags
Downloading https://www.gw-openscience.org/timeline/segments/json/O2_16KHZ_R1/L1_DATA/1186896976/44766/
|======================================================================================================| 123 /123  (100.00%)         0s
Downloading https://www.gw-openscience.org/timeline/segments/json/O2_16KHZ_R1/L1_CBC_CAT2/1186896976/44766/
|======================================================================================================| 127 /127  (100.00%)         0s
Downloading https://www.gw-openscience.org/timeline/segments/json/O2_16KHZ_R1/L1_DATA/1186896976/44766/
|======================================================================================================| 123 /123  (100.00%)         0s
Downloading https://www.gw-openscience.org/timeline/segments/json/O2_16KHZ_R1/L1_NO_CBC_HW_INJ/1186896976/44766/
|======================================================================================================| 132 /132  (100.00%)         0s
Downloading https://www.gw-openscience.org/timeline/segments/json/O2_16KHZ_R1/L1_DATA/1186896976/44766/
|======================================================================================================| 123 /123  (100.00%)         0s
Downloading https://www.gw-openscience.org/timeline/segments/json/O2_16KHZ_R1/L1_NO_BURST_HW_INJ/1186896976/44766/
|======================================================================================================| 134 /134  (100.00%)         0s
2019-07-01 13:27:38,264:INFO : L1: got segments-vetoes flags
Downloading https://www.gw-openscience.org/timeline/segments/json/O2_16KHZ_R1/V1_DATA/1186896976/44766/
|======================================================================================================| 149 /149  (100.00%)         0s
Downloading https://www.gw-openscience.org/timeline/segments/json/O2_16KHZ_R1/V1_CBC_CAT2/1186896976/44766/
|======================================================================================================| 153 /153  (100.00%)         0s
Downloading https://www.gw-openscience.org/timeline/segments/json/O2_16KHZ_R1/V1_DATA/1186896976/44766/
|======================================================================================================| 149 /149  (100.00%)         0s
Downloading https://www.gw-openscience.org/timeline/segments/json/O2_16KHZ_R1/V1_NO_CBC_HW_INJ/1186896976/44766/
|======================================================================================================| 132 /132  (100.00%)         0s
Downloading https://www.gw-openscience.org/timeline/segments/json/O2_16KHZ_R1/V1_DATA/1186896976/44766/
|======================================================================================================| 149 /149  (100.00%)         0s
Downloading https://www.gw-openscience.org/timeline/segments/json/O2_16KHZ_R1/V1_NO_BURST_HW_INJ/1186896976/44766/
|======================================================================================================| 134 /134  (100.00%)         0s
2019-07-01 13:27:42,267:INFO : V1: got segments-vetoes flags
2019-07-01 13:27:42,276:INFO : Entering static psd module.
2019-07-01 13:27:42,277:INFO : Entering template bank generation module.
2019-07-01 13:27:42,277:INFO : Setting template bank from pre-generated bank(s).
2019-07-01 13:27:43,559:INFO : Input bank is a xml.gz file
2019-07-01 13:27:43,559:INFO : convert template bank to HDF
2019-07-01 13:27:43,571:INFO : Leaving template bank generation module.
2019-07-01 13:27:43,571:INFO : Entering split output files module.
2019-07-01 13:27:43,571:INFO : Adding split output file jobs to workflow.
2019-07-01 13:27:43,579:INFO : Leaving split output files module.
2019-07-01 13:27:43,586:INFO : Entering matched-filtering setup module.
2019-07-01 13:27:43,588:INFO : Adding matched-filter jobs to workflow.
2019-07-01 13:27:43,588:INFO : Setting up matched-filtering for V1.
2019-07-01 13:27:43,653:INFO : Setting up matched-filtering for H1.
2019-07-01 13:27:43,719:INFO : Setting up matched-filtering for L1.
2019-07-01 13:27:43,780:INFO : Leaving matched-filtering setup module.
2019-07-01 13:27:43,869:INFO : Setting up coincidence
2019-07-01 13:27:43,907:INFO : ...leaving coincidence 
2019-07-01 13:27:43,907:INFO : No reranking done in this workflow
2019-07-01 13:27:43,907:INFO : Setting up coincidence
2019-07-01 13:27:43,944:INFO : ...leaving coincidence 
2019-07-01 13:27:43,944:INFO : No reranking done in this workflow
2019-07-01 13:27:43,944:INFO : Setting up coincidence
2019-07-01 13:27:43,981:INFO : ...leaving coincidence 
2019-07-01 13:27:43,981:INFO : No reranking done in this workflow
2019-07-01 13:27:43,981:INFO : Setting up coincidence
2019-07-01 13:27:44,180:INFO : ...leaving coincidence 
2019-07-01 13:27:44,180:INFO : No reranking done in this workflow
2019-07-01 13:27:44,181:INFO : Setting up multiifo combine statmap
2019-07-01 13:27:45,368:INFO : Entering minifollowups module
2019-07-01 13:27:45,378:INFO : Leaving minifollowups module
2019-07-01 13:27:45,379:INFO : Entering minifollowups module
2019-07-01 13:27:45,389:INFO : Leaving minifollowups module
2019-07-01 13:27:45,389:INFO : Entering minifollowups module
2019-07-01 13:27:45,400:INFO : Leaving minifollowups module
2019-07-01 13:27:45,400:INFO : Entering minifollowups module
2019-07-01 13:27:45,410:INFO : Leaving minifollowups module
2019-07-01 13:27:45,410:INFO : Entering minifollowups module
2019-07-01 13:27:45,420:INFO : Leaving minifollowups module
2019-07-01 13:27:45,420:INFO : Entering minifollowups module
2019-07-01 13:27:45,430:INFO : Leaving minifollowups module
2019-07-01 13:27:45,430:INFO : Entering minifollowups module
2019-07-01 13:27:45,440:INFO : Leaving minifollowups module
2019-07-01 13:27:45,441:INFO : Entering minifollowups module
2019-07-01 13:27:45,450:INFO : Leaving minifollowups module
2019-07-01 13:27:45,450:INFO : Entering minifollowups module
2019-07-01 13:27:45,460:INFO : Leaving minifollowups module
2019-07-01 13:27:45,460:INFO : Entering minifollowups module
2019-07-01 13:27:45,470:INFO : Leaving minifollowups module
2019-07-01 13:27:45,470:INFO : Entering minifollowups module
2019-07-01 13:27:45,480:INFO : Leaving minifollowups module
2019-07-01 13:27:45,481:INFO : Entering minifollowups module
2019-07-01 13:27:45,491:INFO : Leaving minifollowups module
2019-07-01 13:27:45,491:INFO : Entering minifollowups module
2019-07-01 13:27:45,501:INFO : Leaving minifollowups module
2019-07-01 13:27:54,480:INFO : Entering split output files module.
2019-07-01 13:27:54,480:INFO : Adding split output file jobs to workflow.
2019-07-01 13:27:54,490:INFO : Leaving split output files module.
2019-07-01 13:27:54,490:INFO : Entering injection module.
2019-07-01 13:27:54,507:INFO : Leaving injection module.
2019-07-01 13:27:54,526:INFO : Entering matched-filtering setup module.
2019-07-01 13:27:54,527:INFO : Adding matched-filter jobs to workflow.
2019-07-01 13:27:54,527:INFO : Setting up matched-filtering for V1.
2019-07-01 13:27:54,599:INFO : Setting up matched-filtering for H1.
2019-07-01 13:27:54,664:INFO : Setting up matched-filtering for L1.
2019-07-01 13:27:54,732:INFO : Leaving matched-filtering setup module.
2019-07-01 13:27:54,777:INFO : Setting up coincidence for injections
2019-07-01 13:27:54,864:INFO : ...leaving coincidence for injections
2019-07-01 13:27:54,890:INFO : No reranking done in this workflow
2019-07-01 13:27:54,890:INFO : Setting up coincidence for injections
2019-07-01 13:27:54,971:INFO : ...leaving coincidence for injections
2019-07-01 13:27:54,996:INFO : No reranking done in this workflow
2019-07-01 13:27:54,996:INFO : Setting up coincidence for injections
2019-07-01 13:27:55,078:INFO : ...leaving coincidence for injections
2019-07-01 13:27:55,103:INFO : No reranking done in this workflow
2019-07-01 13:27:55,103:INFO : Setting up coincidence for injections
2019-07-01 13:27:55,480:INFO : ...leaving coincidence for injections
2019-07-01 13:27:55,605:INFO : No reranking done in this workflow
2019-07-01 13:27:55,605:INFO : Setting up multiifo combine statmap
2019-07-01 13:27:55,657:INFO : Entering matched-filtering setup module.
2019-07-01 13:27:55,658:INFO : Adding matched-filter jobs to workflow.
2019-07-01 13:27:55,658:INFO : Setting up matched-filtering for V1.
2019-07-01 13:27:55,807:INFO : Setting up matched-filtering for H1.
2019-07-01 13:27:55,876:INFO : Setting up matched-filtering for L1.
2019-07-01 13:27:55,943:INFO : Leaving matched-filtering setup module.
2019-07-01 13:27:55,989:INFO : Setting up coincidence for injections
2019-07-01 13:27:56,077:INFO : ...leaving coincidence for injections
2019-07-01 13:27:56,105:INFO : No reranking done in this workflow
2019-07-01 13:27:56,105:INFO : Setting up coincidence for injections
2019-07-01 13:27:56,196:INFO : ...leaving coincidence for injections
2019-07-01 13:27:56,223:INFO : No reranking done in this workflow
2019-07-01 13:27:56,223:INFO : Setting up coincidence for injections
2019-07-01 13:27:56,311:INFO : ...leaving coincidence for injections
2019-07-01 13:27:56,337:INFO : No reranking done in this workflow
2019-07-01 13:27:56,337:INFO : Setting up coincidence for injections
2019-07-01 13:27:56,724:INFO : ...leaving coincidence for injections
2019-07-01 13:27:56,850:INFO : No reranking done in this workflow
2019-07-01 13:27:56,850:INFO : Setting up multiifo combine statmap
2019-07-01 13:27:57,201:INFO : Entering injection minifollowups module
2019-07-01 13:27:57,218:INFO : Leaving injection minifollowups module
2019-07-01 13:27:57,508:INFO : Entering injection minifollowups module
2019-07-01 13:27:57,523:INFO : Leaving injection minifollowups module
2019-07-01 13:28:00,041:INFO : Written dax.
```
i.e. "no reranking done in this workflow"

**Comment by ahnitz** [comment]
@tdent @spxiwh This adds the workflow parts needed to do the followup and reranking. I've added a test followup script for now as I'm trying to separate my diverged code path into somewhat easier to manage bits to merge back into master.

**Comment by ahnitz** [comment]
The changes can be described in 3 parts. 
1) Executable to pull out which times to follow up and put into a format the followup code can more easily  manage
2) Script to follow-up these times and generate some statistic (currently included a simple passthrough, more PRs to follow on this point)
3) Script to apply the reranking to the events and produce a modified statmap file
4) Changes to the 2-ifo workflow 
5) Changes to the multi-ifo workflow, the reranking is applied on each detector combo analysis.

**Comment by tdent** [comment]
I doubt I will have much time to look at this this week since trying to get some things finished for Amaldi. See debugging requests ..

File: bin/hdfcoinc/pycbc_apply_rerank [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+
+parser = argparse.ArgumentParser()
+parser.add_argument('--version', action='version',
+    version=pycbc.version.git_verbose_msg)
+parser.add_argument('--verbose', action='store_true')
+parser.add_argument('--stat-files', nargs='+')
+parser.add_argument('--followup-file')
+parser.add_argument('--statmap-file')
+parser.add_argument('--ranking-file')
+parser.add_argument('--output-file')
```
Comment (spxiwh) [inline bin/hdfcoinc/pycbc_apply_rerank]: Please add help messages.

File: bin/hdfcoinc/pycbc_apply_rerank [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+
+args = parser.parse_args()
+pycbc.init_logging(args.verbose)
+
+# Reconstruct the full stat vector
+f = h5py.File(args.followup_file, 'r')
+num = len(f['offsets'])
```
Comment (spxiwh) [inline bin/hdfcoinc/pycbc_apply_rerank]: Can you make the variable name more descriptive? (number of what?)

File: bin/hdfcoinc/pycbc_apply_rerank [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+inv = f['inverse'][:]
```
Comment (spxiwh) [inline bin/hdfcoinc/pycbc_apply_rerank]: I don't know what this is either??

File: bin/hdfcoinc/pycbc_followup_file [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+
+parser = argparse.ArgumentParser()
+parser.add_argument('--version', action='version',
+    version=pycbc.version.git_verbose_msg)
+parser.add_argument('--verbose', action='store_true')
+parser.add_argument('--statmap-file')
+parser.add_argument('--bank-file')
+parser.add_argument('--min-stat', type=float)
+parser.add_argument('--foreground-only', action='store_true')
+parser.add_argument('--output-file')
```
Comment (spxiwh) [inline bin/hdfcoinc/pycbc_followup_file]: Help messages.

File: bin/hdfcoinc/pycbc_apply_rerank [hunk 5]
```diff
@@ -0,0 +1,95 @@
+#!/bin/env python
+"""Rewrite statmap file with a reranking of the statistic values"""
```
Comment (spxiwh) [inline bin/hdfcoinc/pycbc_apply_rerank]: The code doesn't actually seem to do this. From this description I imagine a code where you give it a statmap file and a new statistic (e.g. `new_snr`) and the file gets regenerated. This seems to be doing something else, using this followup_file, which I think needs to be described here.

File: bin/hdfcoinc/pycbc_followup_file [hunk 6]
```diff
@@ -0,0 +1,98 @@
+#!/bin/env python
+"""Followup a candidate"""
```
Comment (spxiwh) [inline bin/hdfcoinc/pycbc_followup_file]: More descriptive please.

File: bin/hdfcoinc/pycbc_rerank_passthrough [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+    version=pycbc.version.git_verbose_msg)
+parser.add_argument('--verbose', action='store_true')
+parser.add_argument('--output-file')
+
+# Options related to getting trigger information from workflow products
+parser.add_argument('--input-file',
+    help="HDF File which gives the trigger followup information for a set")
+parser.add_argument('--start-index', type=int)
+parser.add_argument('--end-index', type=int)
+parser.add_argument('--stride', type=int, default=1)
```
Comment (spxiwh) [inline bin/hdfcoinc/pycbc_rerank_passthrough]: Help messages.

File: bin/workflows/pycbc_create_offline_search_workflow [hunk 8]
```diff
@@ -196,23 +196,18 @@ for i in range(2, len(insps)+1):
...
+                       workflow, bg_file, hdfbank[0], output_dir,
+                       tags=ctagcomb)
 
 combctags = [tag]
-final_bg_file_list = wf.FileList()
-for key in final_bg_files:
-    final_bg_file_list += final_bg_files[key][0][1]
 combined_bg_file = wf.setup_multiifo_combine_statmap(workflow,
-                                                     final_bg_file_list,
+                                                     final_bg_files.values(),
```
Comment (spxiwh) [inline bin/workflows/pycbc_create_offline_search_workflow]: I think this should be a `wf.FileList`, why the change? We should be more careful about not sending lists of Files, but sending FileLists so that the functionality of the latter is available,

File: pycbc/workflow/coincidence.py [hunk 9]
```diff
@@ -141,7 +141,11 @@ def create_node(self, coinc_files, tags=None):
 
         node = Node(self)
         node.set_memory(5000)
-        node.add_input_list_opt('--coinc-files', coinc_files)
```
Comment (spxiwh) [inline pycbc/workflow/coincidence.py]: Let's not add this confusion. Can we not just send a list of one file, if that's the mode its operating in? (Or if we no longer need a list, make this a single file)

File: pycbc/workflow/coincidence.py [hunk 10]
```diff
@@ -174,7 +178,12 @@ def create_node(self, zerolag, full_data,
         node = Node(self)
         node.set_memory(5000)
         node.add_input_list_opt('--zero-lag-coincs', zerolag)
-        node.add_input_list_opt('--full-data-background', full_data)
```
Comment (spxiwh) [inline pycbc/workflow/coincidence.py]: Same comment as above

**Comment by spxiwh** [comment]
Comments mostly related to me being confused about what the executables are actually doing. If this is already described somewhere, I guess just linking to there from in the code would be fine.

File: bin/workflows/pycbc_create_offline_search_workflow (hunk 8 above)
Comment (ahnitz) [reply bin/workflows/pycbc_create_offline_search_workflow]: Reason for the change: It was a bit confusing how there were multiple places carrying around the same information. So it removes a bit of redundancy.  However, you are right that it should probably be a FileList, so I've made it one now.

File: pycbc/workflow/coincidence.py (hunk 9 above)
Comment (ahnitz) [reply pycbc/workflow/coincidence.py]: I'm not sure about this, I'll give it a think...

Comment (ahnitz) [reply pycbc/workflow/coincidence.py]: @spxiwh If possible, I would like to keep this change in.
